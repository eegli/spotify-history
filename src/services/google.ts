import { google } from 'googleapis';
import config from '../config';

const oAuth2Client = new google.auth.OAuth2(
  config.GOOGLE.client_id,
  config.GOOGLE.client_secret
);

oAuth2Client.setCredentials({ refresh_token: config.GOOGLE.refresh_token });

export const drive = google.drive({ version: 'v3', auth: oAuth2Client });
