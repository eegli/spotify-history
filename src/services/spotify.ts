import { initialize, SpotifyClient } from '@spotifly/core';
import config from '../config';
import { DynamoHistoryElement } from '../config/types';

export type HistoryParams = Parameters<
  SpotifyClient['Player']['getRecentlyPlayedTracks']
>[0];

export interface RefreshTokenResponse {
  access_token: string;
  token_type: string;
  scope: string;
  expires_in: number;
}

// Export so we can use it in the tests
export type HistoryResponse = SpotifyApi.UsersRecentlyPlayedTracksResponse;
export type MultipleArtistsResponse = SpotifyApi.MultipleArtistsResponse;

export default class Spotify {
  private client: SpotifyClient;
  private items: SpotifyApi.PlayHistoryObject[] = [];
  // Unix timestamps
  cursorBefore = 0;
  cursorAfter = 0;

  constructor() {
    this.client = initialize({
      clientId: config.SPOTIFY.client_id,
      clientSecret: config.SPOTIFY.client_secret,
      refreshToken: config.SPOTIFY.refresh_token,
    });
  }

  get count() {
    return this.items.length;
  }

  // Gets the raw history from Spotify
  async fetchHistory(params: HistoryParams): Promise<void> {
    const res = await this.client.Player.getRecentlyPlayedTracks({
      ...params,
      limit: 50,
    });

    if (res.data.items.length > 0) {
      this.items = res.data.items;

      const firstElem = this.items[0].played_at;
      const lastElem = this.items[this.items.length - 1].played_at;

      // Manually set the cursors so that we know the time of the
      // first and last song of this history
      this.cursorAfter = new Date(lastElem).getTime();
      this.cursorBefore = new Date(firstElem).getTime();
    }
  }

  // Create the actual history that will be saved in dynamo
  createHistory(): DynamoHistoryElement[] {
    return this.items.reduce((acc, el) => {
      acc.push({
        name: el.track.name,
        id: el.track.id,
        playedAt: new Date(el.played_at).toISOString(),
      });

      return acc;
    }, <DynamoHistoryElement[]>[]);
  }
}
