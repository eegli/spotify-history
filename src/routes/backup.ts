import { googleDrive } from '../services/google';

export type BackupParams<T = unknown> = {
  fileName: string;
  folderName: string;
  meta: Record<string, string | number>;
  data: T;
};

export const backupHistory = async ({
  data,
  meta,
  fileName,
  folderName,
}: BackupParams) => {
  let folderId = '';

  // List all folders that this API client has already created
  const folders = await googleDrive.files.list({
    pageSize: 10,
    // Specify what fields we want to have in the response
    fields: 'files(id, name)',
    q: "mimeType = 'application/vnd.google-apps.folder'",
  });

  // Check if the folder with the given name already exists
  if (folders.data.files?.length) {
    const existingFolderIdx = folders.data.files.findIndex(
      f => f.name === folderName
    );

    if (existingFolderIdx > -1) {
      folderId = folders.data.files[existingFolderIdx].id || '';
    }
  }

  // Folder does not exist, create one
  if (!folderId) {
    const folder = await googleDrive.files.create({
      requestBody: {
        name: folderName,
        mimeType: 'application/vnd.google-apps.folder',
      },
      fields: 'id',
    });

    folderId = folder.data.id || '';
  }

  const body = {
    meta,
    items: data,
  };

  return googleDrive.files.create({
    requestBody: {
      name: fileName + '.json',
      parents: [folderId],
    },
    media: {
      mimeType: 'application/json',
      // Change to JSON.stringify(body) for more compact files
      body: JSON.stringify(body, null, 2),
    },
    fields: 'size,webViewLink',
  });
};
