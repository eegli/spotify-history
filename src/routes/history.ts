import { QueryOptions } from '@aws/dynamodb-data-mapper';
import { AndExpression, ConditionExpression } from '@aws/dynamodb-expressions';
import moment from 'moment';
import { defaults } from '../config';
import { DynamoHistoryElement } from '../config/types';
import { History, HistoryRequired } from '../models/history';
import { dynamoDataMapper } from '../services/dynamo';

const params: ConditionExpression = {
  type: 'Equals',
  subject: 'type',
  object: 'history',
};
export const dynamoGetLatestHistory = async () => {
  const queryOptions: QueryOptions = {
    limit: 1,
    scanIndexForward: false,
  };

  const iterator = dynamoDataMapper.query(History, params, queryOptions);
  for await (const history of iterator) {
    // Return the first element
    return history;
  }
};

export const dynamoSetHistory = async ({
  timestamp,
  songs,
}: HistoryRequired) => {
  const newHistory: History = Object.assign(new History(), {
    timestamp,
    songs,
  });
  return dynamoDataMapper.put(newHistory);
};

export const dynamoGetHistoryRange = async () => {
  // For comparison, the timestamp needs to be an ISO string just like
  // the timestamp from the model
  const timestamp = moment()
    .subtract(...defaults.backupRange)
    .toISOString();

  const dateFilter: ConditionExpression = {
    type: 'GreaterThanOrEqualTo',
    subject: 'timestamp',
    object: timestamp,
  };

  const filters: AndExpression = {
    type: 'And',
    conditions: [params, dateFilter],
  };

  const items: History[] = [];

  const iterator = dynamoDataMapper.query(History, filters);

  for await (const history of iterator) {
    items.push(history);
  }

  // Reduce the history to only get the songs
  return items.reduce((acc, curr) => {
    if (curr.songs) {
      acc.push(...curr.songs);
    }
    return acc;
  }, <DynamoHistoryElement[]>[]);
};
