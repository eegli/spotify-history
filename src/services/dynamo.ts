import config from '../config';

import { DataMapper, QueryOptions } from '@aws/dynamodb-data-mapper';
import { ConditionExpression } from '@aws/dynamodb-expressions';
import DynamoDB from 'aws-sdk/clients/dynamodb';

import { ServiceConfigurationOptions } from 'aws-sdk/lib/service';
import { History, HistoryElement } from '../models/history';

type Options = DynamoDB.DocumentClient.DocumentClientOptions &
  ServiceConfigurationOptions &
  DynamoDB.ClientApiVersions;

const params: Options = {
  region: config.AWS_REGION,
  apiVersion: '2012-08-10',
};

if (process.env.stage === 'local') {
  params.region = 'localhost';
  params.endpoint = 'http://localhost:8000';
}

const mapper = new DataMapper({
  client: new DynamoDB(params),
});

export const dynamoGetLatestHistory = async () => {
  const queryOptions: QueryOptions = {
    limit: 1,
    scanIndexForward: false,
  };

  const params: ConditionExpression = {
    type: 'Equals',
    subject: 'type',
    object: 'history',
  };
  const iterator = mapper.query(History, params, queryOptions);
  for await (const history of iterator) {
    // Return the first element
    return history;
  }
};

export const dynamoSetHistory = async (
  timestamp: string,
  count: number,
  songs: HistoryElement[]
) => {
  const newHistory: History = Object.assign(new History(), {
    timestamp,
    count,
    songs,
  });

  return mapper.put(newHistory);
};
