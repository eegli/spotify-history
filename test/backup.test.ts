import { Context, EventBridgeEvent } from 'aws-lambda';
import { GaxiosPromise, MethodOptions } from 'googleapis-common';
import { drive_v3 } from 'googleapis/build/src/apis/drive/v3';
import { handler } from '../src/handler.backup';
import { backupHistory } from '../src/routes/backup';
import { googleDrive } from '../src/services/google';
import {
  driveCreateResponse,
  driveListResponse,
  dynamoBackupData,
} from './payloads';

// Fake the response from dynamo
jest.mock('../src/routes/history', () => ({
  dynamoGetHistoryRange: jest.fn().mockImplementation(() => dynamoBackupData()),
}));
// Mock the drive client
jest.mock('../src/services/google');

// Setting the right method overloads manually
type DriveListSpy = (
  params?: drive_v3.Params$Resource$Files$List,
  options?: MethodOptions
) => GaxiosPromise<drive_v3.Schema$FileList>;

type DriveCreateSpy = (
  params?: drive_v3.Params$Resource$Files$Create,
  options?: MethodOptions
) => GaxiosPromise<drive_v3.Schema$File>;

const driveListSpy = jest.spyOn(
  googleDrive.files,
  'list'
) as unknown as jest.MockedFunction<DriveListSpy>;

const driveCreateSpy = jest.spyOn(
  googleDrive.files,
  'create'
) as unknown as jest.MockedFunction<DriveCreateSpy>;

/* ____________ */

/* Alternatively to jest.spyOn */

const mockCreate = googleDrive.files
  .create as unknown as jest.MockedFunction<DriveCreateSpy>;
mockCreate.mockResolvedValue(driveCreateResponse);

/* ____________ */

driveListSpy.mockResolvedValue(driveListResponse);
driveCreateSpy.mockResolvedValue(driveCreateResponse);

beforeEach(() => {
  driveListSpy.mockClear();
  driveCreateSpy.mockClear();
});

describe('Backup handler', () => {
  jest.useFakeTimers('modern').setSystemTime(new Date(1996, 3, 21));

  it('creates a backup', async () => {
    await handler(
      {} as EventBridgeEvent<'Scheduled Event', any>,
      {} as Context,
      () => {}
    );
    expect(driveListSpy).toHaveBeenCalledTimes(1);
    expect(driveCreateSpy).toHaveBeenCalledTimes(2);
    expect(driveListSpy.mock.calls[0][0]).toMatchSnapshot();
    expect(driveCreateSpy.mock.calls[0][0]).toMatchSnapshot('creates folder');
    expect(driveCreateSpy.mock.calls[1][0]).toMatchSnapshot('creates file');
  });

  it('does not create a new folder if one is present', async () => {
    const data = { test: true };
    const fileName = 'file';
    const folderName = 'folder';
    const meta = {};

    await backupHistory({ data, fileName, folderName, meta });
    expect(driveCreateSpy).toHaveBeenCalledTimes(1);
  });

  it('returns backup information', async () => {
    const data = { test: true };
    const fileName = 'file';
    const folderName = 'folder';
    const meta = {};

    const res = await backupHistory({ data, fileName, folderName, meta });

    expect(res.data).toHaveProperty('size');
    expect(res.data).toHaveProperty('webViewLink');
  });
});
