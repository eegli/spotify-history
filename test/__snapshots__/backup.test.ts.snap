// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backup handler creates a backup 1`] = `
Object {
  "fields": "files(id, name)",
  "pageSize": 10,
  "q": "mimeType = 'application/vnd.google-apps.folder'",
}
`;

exports[`Backup handler creates a backup: creates file 1`] = `
Object {
  "fields": "size,webViewLink",
  "media": Object {
    "body": "{
  \\"year\\": 1970,
  \\"month\\": 1,
  \\"backup_created\\": \\"1970-01-01T00:00:00.001Z\\",
  \\"count\\": 1,
  \\"items\\": [
    {
      \\"music\\": true,
      \\"tests\\": \\"working!\\"
    }
  ]
}",
    "mimeType": "application/json",
  },
  "requestBody": Object {
    "name": "spotify_bp_1970-00-w1.json",
    "parents": Array [
      "",
    ],
  },
}
`;

exports[`Backup handler creates a backup: creates folder 1`] = `
Object {
  "fields": "id",
  "requestBody": Object {
    "mimeType": "application/vnd.google-apps.folder",
    "name": "SpotifyHistory_test",
  },
}
`;
